pandas_gbq/gbq.py

```diff
@@ -956,61 +803,148 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
+    client = bigquery.Client(project=project_id, credentials=credentials)
+
+    def _set_common_query_settings(query_job):
+        if dialect == 'legacy':
+            query_job.use_legacy_sql = True
+        elif dialect == 'standard':
+            query_job.use_legacy_sql = False
+
+        if configuration:
+            for setting, value in configuration.items():
+                setattr(query_job, setting, value)
+        return query_job
+
+    def sync_query():
+        query_job = client.run_sync_query(query,
+                                          query_parameters=query_parameters)
+        query_job = _set_common_query_settings(query_job)
+        if verbose:
+            print("Query running...")
+        if timeout_ms:
+            query_job.timeout_ms = timeout_ms
+        query_job.run()
+        if not query_job._properties.get("jobComplete", False):
+            raise QueryTimeout("Sync query timed out")
+        if verbose:
+            print("Query done.")
+            if query_job._properties.get("cacheHit", False):
+                print("Cache hit.")
+            else:
+                bytes_billed = int(query_job._properties
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job._properties
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
+                      (sizeof_fmt(bytes_billed), sizeof_fmt(bytes_processed)))
+            print("\nRetrieving results...")
+        return query_job, None
+
+    def async_query():
+        query_job = client.run_async_query(str(uuid.uuid4()),
+                                           query,
+                                           query_parameters=query_parameters)
+        query_job = _set_common_query_settings(query_job)
+        query_job.begin()
+        try:
+            query_results = query_job.results().fetch_data()
+        except:
+            query_results = query_job.result().fetch_data()
+        if verbose:
+            print("Query running...")
+        _wait_for_job(query_job)
+        if verbose:
+            print("Query done.")
+            if query_job._properties["statistics"]["query"].get("cacheHit",
+                                                                False):
+                print("Cache hit.")
+            elif ("statistics" in query_job._properties and
+                    "query" in query_job._properties["statistics"]):
+                bytes_billed = int(query_job
+                                   ._properties["statistics"]["query"]
+                                   .get("totalBytesProcessed", 0))
+                bytes_processed = int(query_job
+                                      ._properties["statistics"]["query"]
+                                      .get("totalBytesBilled", 0))
+                print("Total bytes billed (processed): %s (%s)" %
```

Comment:
why do you have asyc and sync? generally you would *always* do async, and just wait on it for sync.

Reply:
See thread here: https://github.com/pydata/pandas-gbq/pull/25/files/61db19190cdfe690298288cf90e0d62a5b73e95d#r143248553. There's a bug in how async handles timeouts. For now, to provide backwards compatibility for timeouts, we need to do this workaround. When it is fixed, we can then clean this up and remove the sync code. Hope that makes sense, but let know if you think this should be handled differently.

Reply:
see my point above, this should be abstracted away, again to a query.py or something to make this obvious to the reader.